agent/agent.go

```diff
@@ -192,24 +195,49 @@ func (r *agent) LocalStatus() *pb.NodeStatus {
 type dialRPC func(*serf.Member) (*client, error)
 
 // runChecks executes the monitoring tests configured for this agent.
-func (r *agent) runChecks(ctx context.Context) *pb.NodeStatus {
-	var reporter health.Probes
-	// TODO: run tests in parallel
+func (r *agent) runChecks(ctx context.Context) (*pb.NodeStatus, error) {
+	// semaphoreCh limits the number of concurrent checkers
+	semaphoreCh := make(chan struct{}, maxConcurrentCheckers)
+	// channel for collecting resulting health probes
+	probeCh := make(chan health.Probes, len(r.Checkers))
+
 	for _, c := range r.Checkers {
-		log.Infof("running checker %s", c.Name())
-		c.Check(&reporter)
+		select {
+		case semaphoreCh <- struct{}{}:
+			go func(c health.Checker) {
+				var probes health.Probes
+				log.Debugf("running checker %v", c.Name())
+				c.Check(&probes)
+				probeCh <- probes
+				// release checker slot
+				<-semaphoreCh
```

Comment:
could crash or hang during check and consume semaphore slot forever. I would  factor out this goroutine, pass cancel context with timeout and add defer release